content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,208 @@
+---
+title: callPackage, a tool for the lazy
+date: 2022-08-25
+extra:
+  author: Norbert Melzer
+---
+
+The Nix package manager ships with a a special-purpose programming language for creating packages and configurations: the Nix language.
+It is used to build the Nix package collection, known as [`nixpkgs`](https://github.com/nixos/nixpkgs) – the largest, most up-to-date open source software distribution in the world.
+
+Being purely functional, the Nix language allows declaring custom functions to abstract over common patterns.
+One such pattern is parametrization of package definitions, that is, builds which can vary by their dependencies or settings.
+
+`nixpkgs` is a sizeable software project on it's own, with coding conventions and idioms that have emerged over the years.
+It has [established a convention](https://github.com/NixOS/nixpkgs/pull/9869) of composing parameterized packages with automatic settings through a function named [`callPackage`](https://github.com/NixOS/nixpkgs/commit/fd268b4852d39c18e604c584dd49a611dc795a9b).
+
+This article shows how to use it and why it's beneficial.
+
+## Basic examples
+
+Before even discussing the benefits, let's see how it actually gets used.
+
+Given are the files `hello.nix` and `default.nix`:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> {}; in
+pkgs.callPackage ./hello.nix {}
+```
+
+`default.nix` produces a *derivation* from the contents of `hello.nix`.
+This what the Nix package manager calls its build results (which are often packaged executables, but can be arbitrary files), and what the Nix language calls the data type of such an expression.
+
+```nix
+# hello.nix
+{ writeShellScriptBin }:
+writeShellScriptBin "hello" ''
+  echo "hello, world!"
+''
+```
+
+`hello.nix` declares a function which takes as argument an attribute set with one element `writeShellScriptBin`. `writeShellScriptBin` is also a function, which happens to return a derivation. The build result in this case is an executable shell script with the contents `echo "hello world"` named `"hello"`.
+
+Building using `nix-build` (which implicitly evaluates `default.nix` unless told otherwise) will produce `./result/bin/hello`, and running the resulting script will nicely greet you.
+
+As you can see, `writeShellScriptBin` gets passed in by `callPackage` automatically.
+
+It may appear cumbersome to create an extra file for the package in such a simple setup.
+But in fact, this is exactly how `nixpkgs` is organized: every package is a file that declares a function.
+This function takes as arguments the package's dependencies.
```

Comment:
I think explaining how `callPackage` works deserves its own blog post, and fear that just saying "it has stuff inside that it uses implicitly" creates more confusion than it solves, but it's probably better than nothing.

```suggestion
This function takes as arguments Nix expressions which the package requires as dependencies to produce its build outputs.
Conceptually this is an alternative to `import` expressions.
In the case of our `hello.nix` function, we depend on the `writeShellScriptBin` function.
But we could also import other derivations, attribute sets, lists, and whatever else Nix offers.
The difference to `import` is that `callPackages` injects the dependencies explicitly from the context it carries.
We can thus control what gets imported quite easily from within the language.
```

Reply:
Considering how "`import`" (and conceptionally equivalent) usually works in the languages I know beyond nix (python, rust, go, elixir, haskell) I think that "`import` statement" is generally more correct.

```suggestion
This function takes as arguments Nix expressions which the package requires as dependencies to produce its build outputs.
Conceptually this is an alternative to `import` statements.
In the case of our `hello.nix` function, we depend on the `writeShellScriptBin` function.
But we could also import other derivations, attribute sets, lists, and whatever else Nix offers.
The difference to `import` is that `callPackages` injects the dependencies explicitly from the context it carries.
We can thus control what gets imported quite easily from within the language.
```

would be therefore my counterproposal

Reply:
Hm, but `import` is an expression which evaluates whatever is in the given file, with a side effect of reading the file. It does not change the program state. See for example: https://stackoverflow.com/questions/19132/expression-versus-statement/19224#19224

Reply:
An expression can be replaced with its returned value. What value does `import Foo.Bar` return?

It changes as a sideffect the bindings available in the current scope.

`with args; foo` though is an expression, as you could replace that with whatever the value of `args.foo` is.

Also, as far as I remember, rust, haskell, elixir and go all refer to their equivalent of `import` as a statement. Even though indeed in elixir a value is returned, but just having the returned value inline wouldn't compile into the same module.

The only `import` style mechanism that I could live with calling an expression was Cs `#inlcude`, though that doesn't count as it is actually a compiler directive.

Reply:
One more tiny change:

```suggestion
This function takes as arguments Nix expressions which the package requires as dependencies to produce its build outputs.
Conceptually this is an alternative to `import` statements.
In the case of our `hello.nix` function, we depend on the `writeShellScriptBin` function.
But we could also depend on other derivations, attribute sets, lists, and whatever else Nix offers.
The difference to `import` that pulls in dependencies from some global context is that `callPackages` injects the dependencies explicitly from the `pkgs` attribute set where it is defined itself (see file `default.nix`).
We can thus control what gets imported quite easily from within the language.
```

Reply:
> An expression can be replaced with its returned value. What value does import Foo.Bar return?
>
> It changes as a sideffect the bindings available in the current scope, reading a file is not the relevant effect here.

That is news to me. `import` returns the value read from the path in its argument. It does not change anything in its scope. You cannot do `import foo.bar` unless `foo.bar` evaluates to a string that, if interpreted as a file system path, will point to a valid Nix file.

Check the [Nix manual](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-import) for reference.

Reply:
Oh! We are talking about Nix' `import`, totally forgot that one…

Yes, thats an expression.

Reply:
Will need to sit down later to "cherry-pick" from the various suggestions.

In worst case I won't be able to do that before sitting in the train tomorrow.

Reply:
Right, now I understand as well. I was not at all thinking about the nixpkgs `import` expression but about what import means in Python, Haskell etc.... The goal was to refer to something that may be more familiar than Nix to the reader. I see the confusion now.

Reply:
```suggestion
This function takes as arguments Nix expressions which the package requires as dependencies to produce its build outputs.
Conceptually this is an alternative to [`import` expressions](https://nixos.org/manual/nix/stable/language/builtins.html?highlight=import#builtins-import).
In the case of our `hello.nix` function, we depend on the `writeShellScriptBin` function.
But we could also import other derivations, attribute sets, lists, and whatever else `nixpkgs` offers.
The difference to `import` that pulls in dependencies from some context is that `callPackages` injects the dependencies explicitly from the `pkgs` attribute set where it is defined itself (see file `default.nix`).
We can thus control what gets imported quite easily from within the language.
```

This is what I came up with after trying to conclude the discussion here.

Reply:
> The difference to `import` that pulls in dependencies from some context is that `callPackages` injects the dependencies explicitly from the `pkgs` attribute set where it is defined itself (see file `default.nix`).

I don't understand what you mean by "pulls in dependencies from some context". As we discussed above, `import` does nothing except parsing and evaluating a Nix file and returning that value. I think, and you illustrate this later in the article, is that instead of having to explicitly pass each dependency, e.g.

```nix
package = import ./package.nix { inherit dep1, dep2, dep3 };
```

we can use `callPackage` to automatically inject these dependencies from the context we provide it with, which is usually the `pkgs` attrset.

Reply:
Yea I think we are getting tangled up in confusion now. I originally didn't mean to introduce the _Nix_ import expression but I wanted to refer to import statements as they are used in languages like Python to make it easier to understand for readers who come from those worlds. But since Nix also has _imports_ this started the confusion that we are now fighting with. Here is another try:

```suggestion
This function takes as arguments Nix expressions which the package requires to produce its build outputs.
In the case of our `hello.nix` function, we depend on the `writeShellScriptBin` function.
But we could also import other derivations, attribute sets, lists, and whatever else `nixpkgs` offers.
Conceptually this [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) approach is an alternative to `import` statements as they are used in languages like Python that pull in dependencies from some context.
With `callPackages` we can inject dependencies explicitly from a specific `pkgs` attribute set (see file `default.nix`).
This approach gives control over what the dependencies of `hello.nix` are to the caller something which will turn out useful.
```

Reply:
Ah. Okay that would have been super confusing when I read this for the first time without context. We're talking about Nix language, and arguing from the perspective of any other language will lose the rest of the world who are not familiar with any of the specific examples. You just need too much contextual knowledge to make sense of such a side note.

Reply:
Is it still confusing now? I wrote this under the assumption that more people are familiar with Python of Javascript then Nix. There are btw tons of discussions of this dependency-injection / inversion of control pattern in those domains as well.

Reply:
>  readers who do not know Python or Javascript? Which is most people. Even most software developers.

OK among developers I am not so sure about this since Python and JS are language 1 and 2, and if you widen this up to other languages with similar import mechanisms (C, C++, Java) you get surely a vast majority. Even beyond import statements, the most common pattern to depend on another function is rarely to inject them as arguments. Typically you leave control of what to import to the function instead of to the caller which is why this is an _inversion_ of control pattern. This contrast with the usual approach, the benefits and disadvantages that it brings, is what makes the pattern interesting and noteworthy in the first place and difficult to understand. If `hello.nix` would just import `writeShellScriptBin` from another file or some global context of nixpkgs we would probably not have started writing this article.

But I'm fine leaving it out if it is so controversial. I think the post doesn't depend on it and we can agree to disagree :) Here is a version without it:

```
This function takes as arguments Nix expressions which the package requires to produce its build outputs.
In the case of our `hello.nix` function, we depend on the `writeShellScriptBin` function.
But we could also depend on other derivations, attribute sets, lists, and whatever else `nixpkgs` offers.
With `callPackages` we can inject these dependencies explicitly from a specific `pkgs` attribute set (see file `default.nix`).
This approach gives control over the dependencies of `hello.nix` to the caller something which will turn out useful.
```

Reply:
> With `callPackages` we can inject these dependencies explicitly from a specific `pkgs` attribute set (see file `default.nix`).

Shouldn't this be "implicitely"? As `callPackage` will use what it "knows" unless explicitely overriden in the second argument?

I like this last version though. It manages to not mention `import` at all quite well and avoids the confusion we had.